File: src_tf/demo_cpp_tf/src/dynamic_tf_broadcaster.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/transform_stamped.hpp> //提供消息接口
#include <tf2/LinearMath/Quaternion.h> //四元数类
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp> //四元数与消息的转换
#include <tf2_ros/transform_broadcaster.h> //坐标容器
#include <chrono>

using namespace std::chrono_literals;


class TFBroadcasterNode : public rclcpp::Node
{
private:
    std::shared_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
    rclcpp::TimerBase::SharedPtr timer_;
public:
    TFBroadcasterNode() : Node("tf_broadcaster")
    {
        this->tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
        this->timer_ = this->create_wall_timer(
            100ms,
            std::bind(&TFBroadcasterNode::publish_tf, this)
        );
    }

    void publish_tf()
    {
        geometry_msgs::msg::TransformStamped transform;

        transform.header.stamp = this->get_clock()->now();
        transform.header.frame_id = "map";
        transform.child_frame_id = "base_link";

        transform.transform.translation.x = 2.0;
        transform.transform.translation.y = 3.0;
        transform.transform.translation.z = 0.0;

        tf2::Quaternion q;
        q.setRPY(0.0, 0.0, 30*M_PI/180.0); // Roll, Pitch, Yaw
        transform.transform.rotation = tf2::toMsg(q);

        this->tf_broadcaster_->sendTransform(transform);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TFBroadcasterNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src_tf/demo_python_tf/demo_python_tf/dynamic_tf_broadcaster.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformBroadcaster #坐标发布器
from geometry_msgs.msg import TransformStamped #消息接口
from tf_transformations import quaternion_from_euler #欧拉角转四元数
import math 

class TFBroadCaster(Node):
    def __init__(self):
        super().__init__('tf_broadcaster')
        self.broadcaster_ = TransformBroadcaster(self)
        self.timer_ = self.create_timer(0.01, self.publish_tf) #定时器周期性发布tf

    #发布tf
    def publish_tf(self):
        #从camera_link到bottle_link的坐标关系
        transform = TransformStamped()

        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = 'camera_link' #父坐标系名称
        transform.child_frame_id = 'bottle_link' #子坐标系名称

        transform.transform.translation.x = 0.2
        transform.transform.translation.y = 0.3
        transform.transform.translation.z = 0.5

        #欧拉角转四元数 返回值为元组 xyzw
        q = quaternion_from_euler(0, 0, 0) #math用于角度转弧度
        transform.transform.rotation.x = q[0]
        transform.transform.rotation.y = q[1]
        transform.transform.rotation.z = q[2]
        transform.transform.rotation.w = q[3]

        #发布静态坐标变换
        self.broadcaster_.sendTransform(transform)
        self.get_logger().info(f'{transform}')
def main():
    rclpy.init()
    node = TFBroadCaster()
    rclpy.spin(node)
    rclpy.shutdown()
